velox/functions/sparksql/aggregates/ModeAggregate.cpp

```diff
@@ -0,0 +1,580 @@
+
+    static constexpr bool is_fixed_size_ = false;
+
+    void addInput(HashStringAllocator* allocator, arg_type<Varchar> data) {
+      values[store(data, allocator)]++;
+    }
+
+    void combine(
+        HashStringAllocator* allocator,
+        arg_type<IntermediateType> other) {
+      for (const auto& [key, count] : other) {
+        if (count.has_value()) {
+          values[store(key, allocator)] += count.value();
+        }
+      }
+    }
+
+    bool writeIntermediateResult(out_type<IntermediateType>& out) {
+      out.reserve(values.size());
+      for (const auto& [value, count] : values) {
+        out.add_item() = std::make_tuple(value, count);
+      }
+      return true;
+    }
+
+    bool writeFinalResult(out_type<OutputType>& out) {
+      auto iterator = std::max_element(
+          std::begin(values),
+          std::end(values),
+          [](const std::pair<StringView, int64_t>& p1,
+             const std::pair<StringView, int64_t>& p2) {
+            return p1.second < p2.second;
+          });
+      out = iterator->first;
+      return true;
+    }
+
+    void destroy(HashStringAllocator* allocator) {
+      strings.free(*allocator);
+      using TValues = decltype(values);
+      values.~TValues();
+    }
+
+   private:
+    // Stores the non-inlined string in memory blocks managed by
+    // HashStringAllocator, if it's not exist in the values map.
+    StringView store(StringView value, HashStringAllocator* allocator) {
+      if (!value.isInline()) {
+        auto it = values.find(value);
+        if (it == values.end()) {
+          value = strings.append(value, *allocator);
+        }
+      }
+      return value;
+    }
+  };
+};
+
+struct ComplexTypeAccumulator {
+  using ValueMap = folly::F14FastMap<
+      AddressableNonNullValueList::Entry,
+      int64_t,
+      AddressableNonNullValueList::Hash,
+      AddressableNonNullValueList::EqualTo,
+      AlignedStlAllocator<
+          std::pair<const AddressableNonNullValueList::Entry, int64_t>,
+          16>>;
+
+  ValueMap values;
+
+  AddressableNonNullValueList serializedValues;
+
+  ComplexTypeAccumulator(const TypePtr& type, HashStringAllocator* allocator)
+      : values{
+            0,
+            AddressableNonNullValueList::Hash{},
+            AddressableNonNullValueList::EqualTo{type},
+            AlignedStlAllocator<
+                std::pair<const AddressableNonNullValueList::Entry, int64_t>,
+                16>(allocator)} {}
+
+  size_t size() const {
+    return values.size();
+  }
+
+  void addValue(
+      DecodedVector& decoded,
+      vector_size_t index,
+      HashStringAllocator* allocator) {
+    addValueWithCount(*decoded.base(), decoded.index(index), 1, allocator);
+  }
+
+  void addValueWithCount(
+      const BaseVector& keys,
+      vector_size_t index,
+      int64_t count,
+      HashStringAllocator* allocator) {
+    auto entry = serializedValues.append(keys, index, allocator);
+
+    auto it = values.find(entry);
```

Comment:
It seems like we are accumulating all distinct values with counts while we only need to keep track of distinct values with top count.

Reply:
Yes, the final result is the the value with top count.